File: src/main/java/de/tectoast/emolga/utils/TypicalSets.java
package de.tectoast.emolga.utils;

import de.tectoast.emolga.commands.Command;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import org.jsolf.JSONObject;

import java.awt.*;
import java.util.Collection;
import java.util.Optional;
import java.util.stream.Collectors;

import static de.tectoast.emolga.commands.Command.getGerNameNoCheck;

public record TypicalSets(JSONObject json) {
    private static TypicalSets instance;

    public static void init(JSONObject o) {
        if (instance == null) instance = new TypicalSets(o);
    }

    public static TypicalSets getInstance() {
        return instance;
    }

    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
    public void add(String pokemon, Collection<String> movesList, Optional<String> item, Optional<String> ability) {
        JSONObject mon = json.createOrGetJSON(pokemon);
        mon.increment("uses");
        movesList.forEach(m -> mon.createOrGetJSON("moves").increment(getGerNameNoCheck(m)));
        item.ifPresent(s -> mon.createOrGetJSON("items").increment(getGerNameNoCheck(s)));
        ability.ifPresent(s -> mon.createOrGetJSON("abilities").increment(getGerNameNoCheck(s)));
    }

    public synchronized void save() {
        Command.save(json, "typicalsets.json");
    }

    public MessageEmbed buildPokemon(String pokemon) {
        if (!json.has(pokemon)) {
            return new EmbedBuilder().setTitle("Dieses Pokemon ist noch nicht in den TypicalSets erfasst!").setColor(Color.RED).build();
        }
        JSONObject mon = json.getJSONObject(pokemon);
        double uses = mon.getInt("uses");
        return new EmbedBuilder().addField("Attacken", mon.optJSONObject("moves", new JSONObject()).toMap().entrySet().stream().map(e -> {
            String usesStr = String.valueOf((int) e.getValue() / uses * 100f);
            return e.getKey() + ": " + usesStr.substring(0, Math.min(usesStr.length(), 5)) + "%";
        }).limit(5).collect(Collectors.joining("\n")), true).addField("Items", mon.optJSONObject("items", new JSONObject()).toMap().entrySet().stream().map(e -> {
            String itemsStr = String.valueOf((int) e.getValue() / uses * 100f);
            return e.getKey() + ": " + itemsStr.substring(0, Math.min(itemsStr.length(), 5)) + "%";
        }).limit(5).collect(Collectors.joining("\n")), true).addField("Fähigkeiten", mon.optJSONObject("abilities", new JSONObject()).toMap().entrySet().stream().map(e -> {
            String abilitiesStr = String.valueOf((int) e.getValue() / uses * 100f);
            return e.getKey() + ": " + abilitiesStr.substring(0, Math.min(abilitiesStr.length(), 5)) + "%";
        }).limit(5).collect(Collectors.joining("\n")), true).setColor(Color.CYAN).setTitle(pokemon).build();
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/records/TypicalSets.kt
package de.tectoast.emolga.utils.records

import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.Command.Companion.getGerNameNoCheck
import de.tectoast.emolga.commands.Command.Companion.save
import de.tectoast.emolga.commands.embedColor
import de.tectoast.emolga.commands.increment
import dev.minn.jda.ktx.messages.Embed
import kotlinx.serialization.Serializable
import net.dv8tion.jda.api.entities.MessageEmbed
import kotlin.math.roundToInt

object TypicalSets {
    val json: MutableMap<String, TypicalSet> =
        Command.load<Map<String, TypicalSet>>("./typicalsets.json").toMutableMap()

    fun add(pokemon: String, movesList: Collection<String>, item: String?, ability: String?) {
        val mon = json.getOrPut(pokemon) { TypicalSet(0, mutableMapOf(), mutableMapOf(), mutableMapOf()) }
        mon.uses++
        movesList.forEach {
            mon.moves.increment(getGerNameNoCheck(it))
        }
        item?.let {
            mon.items.increment(getGerNameNoCheck(it))
        }
        ability?.let {
            mon.abilities.increment(getGerNameNoCheck(it))
        }
    }

    @Synchronized
    fun save() {
        save(json, "typicalsets.json")
    }

    private val comp = compareByDescending<Map.Entry<String, Int>> { it.value }

    private fun MutableMap<String, Int>.max5(uses: Double) = entries.sortedWith(comp).take(5)
        .joinToString("\n") { "${it.key}: ${(it.value.toDouble() / uses * 10000.0).roundToInt() / 100.0}%" }

    fun buildPokemon(pokemon: String): MessageEmbed {
        val mon = json[pokemon] ?: return Embed(
            title = "Dieses Pokemon ist noch nicht in den TypicalSets erfasst!",
            color = 0xFF0000
        )
        val uses = mon.uses.toDouble()
        return Embed {
            title = "TypicalSets für $pokemon"
            field {
                name = "Attacken"
                value = mon.moves.max5(uses)
                inline = true
            }
            field {
                name = "Items"
                value = mon.items.max5(uses)
                inline = true
            }
            field {
                name = "Fähigkeiten"
                value = mon.abilities.max5(uses)
                inline = true
            }
            color = embedColor
        }
    }
}

@Serializable
data class TypicalSet(
    var uses: Int = 0,
    val moves: MutableMap<String, Int> = mutableMapOf(),
    val items: MutableMap<String, Int> = mutableMapOf(),
    val abilities: MutableMap<String, Int> = mutableMapOf()
)
